File: Python/time_zone_offset.py
import pytz
from datetime import datetime

def time_zone_offset(tz_list, datetime_list=None):
    """
    Calculate Time Zone Offset
    
    This function calculates the offset in hours and minutes of a given time zone or list of time zones from UTC.
    It returns None for invalid time zones and prints a warning.
    
    Parameters:
        tz_list (list): A list of strings representing the time zone(s) for which to calculate the offset.
        datetime_list (list, optional): A list of datetime objects representing the datetime(s) for which to calculate the offset.
                                         Default is the current system time for each time zone in tz_list.
                                         
    Returns:
        list: A list of strings representing the offset(s) from UTC in the format "HH:MM".
        
    Examples:
        >>> time_zone_offset(["America/New_York", "Europe/London", "Asia/Tokyo"])
    """
    if datetime_list is None:
        datetime_list = [datetime.now(pytz.timezone(tz)) for tz in tz_list]
    
    if len(datetime_list) != len(tz_list):
        raise ValueError("Length of datetime_list and tz_list must be the same.")
    
    offsets = []
    for tz, dt in zip(tz_list, datetime_list):
        if tz not in pytz.all_timezones:
            print(f"Warning: Invalid time zone detected: {tz}. Returning None for this entry.")
            offsets.append(None)
            continue
        
        localized_dt = pytz.timezone(tz).localize(dt)
        offset_minutes = localized_dt.utcoffset().total_seconds() / 60
        offset_hours = int(offset_minutes // 60)
        offset_minutes = int(offset_minutes % 60)
        offsets.append(f"{offset_hours:+03d}:{offset_minutes:02d}")
        
    return offsets

# Example usage:
tz_list = ["America/New_York", "Europe/London", "Asia/Tokyo", "Invalid/TimeZone"]
datetime_list = [datetime.now() for _ in tz_list]
offsets = time_zone_offset(tz_list, datetime_list)
print(offsets)

File: R/time_zone_offset.R
# Load the necessary library
library(lubridate)

# Define the enhanced time_zone_offset function
time_zone_offset <- function(tz, datetime = Sys.time()) {
  #' Calculate Time Zone Offset
  #'
  #' This function calculates the offset in hours and minutes of a given time zone or vector of time zones from UTC.
  #' It returns NA for invalid time zones and emits a warning.
  #'
  #' @param tz A character vector representing the time zone(s) for which to calculate the offset.
  #' @param datetime A POSIXct object or vector representing the datetime(s) for which to calculate the offset. Default is the current system time.
  #' @return A character vector representing the offset(s) from UTC in the format "HH:MM".
  #' @examples
  #' time_zone_offset(c("America/New_York", "Europe/London", "Asia/Tokyo"))
  
  # If tz is not a character vector, or if it has zero length, return an error
  if (!is.character(tz) || length(tz) == 0) {
    stop("Invalid time zone provided. tz should be a non-empty character vector.")
  }
  
  # Handle vectorized datetime input
  if (length(datetime) != length(tz) && length(datetime) > 1) {
    stop("Length of datetime vector and tz vector must be the same.")
  }
  
  # Validate each time zone and give a warning for invalid ones
  valid_tz <- tz %in% OlsonNames()
  if (any(!valid_tz)) {
    warning("Invalid time zones detected. Returning NA for these entries.")
  }
  
  # Vectorized computation of time zone offset
  offsets <- sapply(seq_along(tz), function(i) {
    zone <- tz[i]
    
    # If the time zone is not valid, return NA
    if (!valid_tz[i]) {
      return(NA_character_)
    }
    
    # Handle NA datetime
    if (is.na(datetime[i])) {
      return(NA_character_)
    }
    
    # Get the time in the specified time zone
    current_time <- with_tz(datetime[i], zone)
    
    # Calculate the offset in minutes
    offset_minutes <- as.numeric(format(current_time, "%z"))
    
    # Convert the offset to hours and minutes
    offset_hours <- offset_minutes %/% 100
    offset_minutes <- offset_minutes %% 100
    
    # Return the offset as a string in the format "HH:MM"
    return(sprintf("%+03d:%02d", offset_hours, offset_minutes))
  })
  
  return(offsets)
}

# Example usage:
tz_vector <- c("America/New_York", "Europe/London", "Asia/Tokyo", "Invalid/TimeZone")
datetime_vector <- c(Sys.time(), Sys.time(), Sys.time(), Sys.time())
offsets <- time_zone_offset(tz_vector, datetime_vector)
print(offsets)
